// testcafe/lib/test-run/execute-js-expression.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const vm_1 = require("vm");
const module_1 = __importDefault(require("module"));
const path_1 = require("path");
const nanoid_1 = __importDefault(require("nanoid"));
const test_run_1 = require("../errors/test-run");
const exportable_lib_1 = __importDefault(require("../api/exportable-lib"));
const ERROR_LINE_COLUMN_REGEXP = /:(\d+):(\d+)/;
const ERROR_LINE_OFFSET = -1;
const ERROR_COLUMN_OFFSET = -4;
const contexts = {};
// NOTE: do not beautify this code since offsets for error lines and columns are coded here
function wrapInAsync(expression) {
    return '(async function() {\n' +
        expression + ';\n' +
        '});';
}
function getErrorLineColumn(err) {
    const result = err.stack.match(ERROR_LINE_COLUMN_REGEXP);
    const line = parseInt(result[1], 10);
    const column = parseInt(result[2], 10);
    return { line, column };
}
function formatExpression(expression) {
    const expresionMessage = expression.split('\n');
    return '[JS code]\n' + expresionMessage.map(str => {
        return ' '.repeat(10) + str;
    }).join('\n');
}
function createRequire(filename) {
    if (module_1.default.createRequireFromPath)
        return module_1.default.createRequireFromPath(filename);
    const dummyModule = new module_1.default(filename, module);
    dummyModule.filename = filename;
    dummyModule.paths = [filename].concat(module.paths);
    return id => dummyModule.require(id);
}
function createSelectorDefinition(testRun) {
    return (fn, options = {}) => {
        const { skipVisibilityCheck, collectionMode } = contexts[testRun.contextId].options;
        if (skipVisibilityCheck)
            options.visibilityCheck = false;
        if (testRun && testRun.id)
            options.boundTestRun = testRun;
        if (collectionMode)
            options.collectionMode = collectionMode;
        return exportable_lib_1.default.Selector(fn, options);
    };
}
function createClientFunctionDefinition(testRun) {
    return (fn, options = {}) => {
        if (testRun && testRun.id)
            options.boundTestRun = testRun;
        return exportable_lib_1.default.ClientFunction(fn, options);
    };
}
function getExecutingContext(testRun, options = {}) {
    const contextId = testRun.contextId || nanoid_1.default(7);
    if (!contexts[contextId])
        contexts[contextId] = createExecutingContext(testRun);
    contexts[contextId].options = options;
    testRun.contextId = contextId;
    return contexts[contextId];
}
function createExecutingContext(testRun) {
    const filename = testRun.test.testFile.filename;
    const replacers = {
        require: createRequire(filename),
        __filename: filename,
        __dirname: path_1.dirname(filename),
        t: testRun.controller,
        Selector: createSelectorDefinition(testRun),
        ClientFunction: createClientFunctionDefinition(testRun),
        Role: exportable_lib_1.default.Role,
        RequestLogger: exportable_lib_1.default.RequestLogger,
        RequestMock: exportable_lib_1.default.RequestMock,
        RequestHook: exportable_lib_1.default.RequestHook
    };
    return vm_1.createContext(new Proxy(replacers, {
        get: (target, property) => {
            if (replacers.hasOwnProperty(property))
                return replacers[property];
            if (global.hasOwnProperty(property))
                return global[property];
            throw new Error(`${property} is not defined`);
        }
    }));
}
function createErrorFormattingOptions(expression) {
    return {
        filename: formatExpression(expression),
        lineOffset: ERROR_LINE_OFFSET,
        columnOffset: ERROR_COLUMN_OFFSET
    };
}
function executeJsExpression(expression, testRun, options) {
    const context = getExecutingContext(testRun, options);
    const errorOptions = createErrorFormattingOptions(expression);
    return vm_1.runInContext(expression, context, errorOptions);
}
exports.executeJsExpression = executeJsExpression;
async function executeAsyncJsExpression(expression, testRun, callsite) {
    const context = getExecutingContext(testRun);
    const errorOptions = createErrorFormattingOptions(expression);
    try {
        return await vm_1.runInContext(wrapInAsync(expression), context, errorOptions)();
    }
    catch (err) {
        if (err.isTestCafeError)
            throw err;
        const { line, column } = getErrorLineColumn(err);
        throw new test_run_1.ExecuteAsyncExpressionError(err, expression, line, column, callsite);
    }
}
exports.executeAsyncJsExpression = executeAsyncJsExpression;
